src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -169,6 +170,26 @@ private Method findSetsFactory(Method testMethod, String factoryMethodName) {
 		return factory;
 	}
 
+	private String getFactoryMethodName(String factoryMethodName) {
+		if (factoryMethodName.contains("("))
+			factoryMethodName = factoryMethodName.substring(0, factoryMethodName.indexOf('('));
+		if (factoryMethodName.contains("#"))
+			return factoryMethodName.substring(factoryMethodName.indexOf('#') + 1);
+		return factoryMethodName;
+	}
+
+	private Class<?> getExplicitOrImplicitClass(Method testMethod, String factoryMethodName) {
+		if (factoryMethodName.contains("#")) {
+			String className = factoryMethodName.substring(0, factoryMethodName.indexOf('#'));
+			return ReflectionSupport
+					.tryToLoadClass(className)
+					.getOrThrow(ex -> new ExtensionConfigurationException(
+						format("Class %s not found, referenced in method %s", className, testMethod.getName()), ex));
+
+		}
+		return testMethod.getDeclaringClass();
+	}
```

Comment:
Is there no tooling we can use instead of string manipulation?

Reply:
Do you have a suggestion? I just copy-pasted from JUnit's `@MethodSource`.

Reply:
Nope, it was an honest (open) question. String manipulation just feels so yucky.

Maybe there's an OSS project hiding here - either already existing or desperately needed. 😉